Separate event store from event stream

tofarr opened the pull request:
- [ ] This change is worth documenting at https://docs.all-hands.dev/
- [ ] Include this change in the Release Notes. If checked, you **must** provide an **end-user friendly** description for your change below

**End-user friendly description of the problem this fixes or functionality that this introduces.**
This PR separates the concerns of threading and coordination from the concerns of event storage - a new class `StoredEventList` is introduced to this end, and EventStream now extends this.

---
**Give a summary of what the PR does, explaining any non-trivial design decisions.**
The event stream is designed to coordinate read / write access for events in a conversation. This is overkill for the return values from the conversation manager - we just want to expose a way to read the existing events for a conversation in it's return values. (This extra flexibility is required in the SAAS env where clustering is in play, and the conversation may be running remotely in a different worker / process).

---
**Link of any specific issues this addresses.**

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:90839be-nikolaik   --name openhands-app-90839be   docker.all-hands.dev/all-hands-ai/openhands:90839be
```

rbren commented on openhands/events/stored_event_list.py:
```diff
@@ -0,0 +1,238 @@
+
+        try:
+            events_dir = get_conversation_events_dir(self.sid, self.user_id)
+            events += self.file_store.list(events_dir)
```
why `+=` since it's empty? probably more efficent to just to `=`

tofarr replied:
I noticed that too - it was copied from the previous code. The part that really concerned me was L42 where we add a second list - I would have thought this should be a replace operation, but I didn't want to mess with the logic.

rbren commented on openhands/server/conversation_manager/standalone_conversation_manager.py:
```diff
@@ -300,19 +301,19 @@ async def maybe_start_agent_loop(
+        return event_store
 
-    async def _get_event_stream(
+    async def _get_event_store(
```
this still returns `agent_session.event_stream`

tofarr replied:
EventStream extends StoredEventList, so the contract for the method is valid. I suppose somebody could force the type hint back to EventStream, or ignore the type hints altogether, but the linter would complain.

rbren reviewed (commented):
I think we can stop returning EventStreams and EventStores from the conversation manager, and just construct EventStore directly from convo IDs! Much simpler

enyst commented on openhands/events/stored_event_list.py:
```diff
@@ -0,0 +1,238 @@
+@dataclass
+class StoredEventList:
+    """
+    A stored list of events backing a conversation
```
Totally nitpicking, what do you think, maybe we could call this `StoredEventStream`? Just thinking that nothing in its interface or structure uses a list, it's still streaming.

tofarr replied:
Maybe "StoredEvents"?

rbren replied:
hmm yeah `Stream` I think of as more of a live-updating thing, whereas this is a static list

enyst replied:
I see we made it `EventStore`, solved!

tofarr commented:
@enyst / @rbren - I've made some updates here...
* Renamed `StoredEventList` -> `EventStore` (Following the pattern of SettingsStore, ConversationStore, etc...)
* Changed the EventStore class so that it is possible to create instances without recalculating the cur_id - so we can create these easily and cheaply. (The default_value is -1 and we use the dataclass post_init to fix this if it is < 0)
* The conversation manager now returns an EventStore instance, so it is not possible for somebody to accidentally cast it back to a stream.

enyst commented on openhands/events/stream.py:
```diff
@@ -90,40 +65,11 @@ def __init__(self, sid: str, file_store: FileStore, user_id: str | None = None):
         self.secrets = {}
 
         # load the stream
         self.__post_init__()
```
Nit: do we still need this? It's here because `EventStream` is not a dataclass, and we wanted it called. Now post init moved to `EventStore` which is a dataclass, so I think it's called implicitly even if the subclass is instantiated?

tofarr replied:
Good catch - I removed that (Also changed the name to `cur_id`)

tofarr replied:
My bad (Missed a push)

enyst reviewed (approved):
Looks cleaner!
